### PR TITLE
pkg: fix cross compiling with cmake on macOS

### PR DESCRIPTION
**Pull request by smlng**
### Contribution description

This PR fixes cross compiling on macOS for packages that use `cmake`, these currently are ccn-lite and relic. While relic had this already I recently ran in troubles with ccn-lite, which is fixed here.

For future use by other packages the helper script is moved (and adapted a bit) to the tools directory, and Makefiles for both packages are adapted accordingly.

### CODE REVIEW

File: dist/tools/cmake/generate-xcompile-toolchain.pl [hunk 1]
```diff
@@ -1,26 +1,22 @@
 #!/usr/bin/env perl
-print "INCLUDE(CMakeForceCompiler)\n";
-print "\n";
-print "\n";
 print "SET(CMAKE_SYSTEM_NAME Generic)\n";
 print "SET(CMAKE_SYSTEM_VERSION 1)\n";
 print "\n";
+print "# specify the cross compiler\n";
+#print "SET(CMAKE_AR \"$ENV{AR}\" CACHE STRING \"\")\n";
```
Comment (kaspar030) [inline dist/tools/cmake/generate-xcompile-toolchain.pl]: here is something commented out

File: pkg/ccn-lite/Makefile [hunk 2]
```diff
@@ -7,8 +7,22 @@ PKG_LICENSE=ISC
 
 export RIOT_CFLAGS = ${CFLAGS} ${INCLUDES}
 
-all: git-download
-	cd $(PKG_BUILDDIR)/src && cmake -DCCNL_RIOT=1 -DRIOT_CFLAGS="${RIOT_CFLAGS}" . && make
+$(info $(PKG_BUILDDIR))
```
Comment (kaspar030) [inline pkg/ccn-lite/Makefile]: debug output?

Comment (smlng) [reply pkg/ccn-lite/Makefile]: yep, forgot to remove

File: dist/tools/cmake/generate-xcompile-toolchain.pl (hunk 1 above)
Comment (smlng) [reply dist/tools/cmake/generate-xcompile-toolchain.pl]: not needed, will remove

**Comment by smlng** [comment]
@kaspar030 fixed, amended directly

File: dist/tools/cmake/generate-xcompile-toolchain.pl [hunk 3]
```diff
@@ -1,26 +1,21 @@
 #!/usr/bin/env perl
```
Comment (cladmi) [inline dist/tools/cmake/generate-xcompile-toolchain.pl]: As moving to a general RIOT script, I would prefer to have it in either `sh/bash` or `python` than `perl`.

File: makefiles/vars.inc.mk [hunk 4]
```diff
@@ -21,6 +21,7 @@ export RIOTBASE              # The root folder of RIOT. The folder where this ve
 export RIOTCPU               # For third party CPUs this folder is the base of the CPUs.
 export RIOTBOARD             # For third party BOARDs this folder is the base of the BOARDs.
 export RIOTPKG               # For overriding RIOT's pkg directory
+export RIOTTOOLS             # For overriding RIOT's tools directory
```
Comment (cladmi) [inline makefiles/vars.inc.mk]: Could this be done in a separate PR/issue, because right now there are still a lot of `dist/tools` in the repository, and only two packages using the variable does not help keeping things consistent:
```
git grep 'dist/tools' | wc -l
219
```

File: pkg/ccn-lite/Makefile [hunk 5]
```diff
@@ -7,8 +7,17 @@ PKG_LICENSE=ISC
 
 export RIOT_CFLAGS = ${CFLAGS} ${INCLUDES}
 
-all: git-download
-	cd $(PKG_BUILDDIR)/src && cmake -DCCNL_RIOT=1 -DRIOT_CFLAGS="${RIOT_CFLAGS}" . && make
+all: $(PKG_BUILDDIR)/src/Makefile
+	$(MAKE) -C $(PKG_BUILDDIR)/src && \
 	cp $(PKG_BUILDDIR)/src/lib/libccnl-riot.a ${BINDIR}/ccn-lite.a
 
+$(PKG_BUILDDIR)/src/Makefile: $(PKG_BUILDDIR)/xcompile-toolchain.cmake
```
Comment (cladmi) [inline pkg/ccn-lite/Makefile]: Please use a variable for `$(PKG_BUILDDIR)/xcompile-toolchain.cmake` its repeated 3 times in the build.

File: pkg/ccn-lite/Makefile [hunk 6]
```diff
@@ -7,8 +7,17 @@ PKG_LICENSE=ISC
 
 export RIOT_CFLAGS = ${CFLAGS} ${INCLUDES}
 
-all: git-download
-	cd $(PKG_BUILDDIR)/src && cmake -DCCNL_RIOT=1 -DRIOT_CFLAGS="${RIOT_CFLAGS}" . && make
+all: $(PKG_BUILDDIR)/src/Makefile
+	$(MAKE) -C $(PKG_BUILDDIR)/src && \
```
Comment (cladmi) [inline pkg/ccn-lite/Makefile]: `&&` not required here.

File: pkg/ccn-lite/Makefile [hunk 7]
```diff
@@ -7,8 +7,17 @@ PKG_LICENSE=ISC
...
 	cp $(PKG_BUILDDIR)/src/lib/libccnl-riot.a ${BINDIR}/ccn-lite.a
 
+$(PKG_BUILDDIR)/src/Makefile: $(PKG_BUILDDIR)/xcompile-toolchain.cmake
+	cd $(PKG_BUILDDIR)/src && \
+	cmake -DCMAKE_TOOLCHAIN_FILE=$(PKG_BUILDDIR)/xcompile-toolchain.cmake \
+		  -DCCNL_RIOT=1 -DRIOT_CFLAGS="${RIOT_CFLAGS}" .
```
Comment (cladmi) [inline pkg/ccn-lite/Makefile]: You could remove the `cd` by using `$(@D)` instead of `.`

File: pkg/ccn-lite/Makefile [hunk 8]
```diff
@@ -7,8 +7,17 @@ PKG_LICENSE=ISC
...
+
+$(PKG_BUILDDIR)/xcompile-toolchain.cmake: git-download
+	cd $(PKG_BUILDDIR) && \
+	perl $(RIOTTOOLS)/cmake/generate-xcompile-toolchain.pl > xcompile-toolchain.cmake
```
Comment (cladmi) [inline pkg/ccn-lite/Makefile]: Here too, you could just use `$@` instead of cd.

File: pkg/relic/Makefile [hunk 9]
```diff
@@ -5,22 +5,27 @@ PKG_LICENSE=LGPL-2.1
 
 .PHONY: all
 
+CFLAGS += -Wno-gnu-zero-variadic-macro-arguments -Wno-unused-function -Wno-newline-eof
+
 all: $(PKG_BUILDDIR)/Makefile
-	"$(MAKE)" -C $(PKG_BUILDDIR) && \
+	$(MAKE) -C $(PKG_BUILDDIR) && \
```
Comment (cladmi) [inline pkg/relic/Makefile]: Removed the `"` around MAKE, also `&&` not required.

File: pkg/relic/Makefile [hunk 10]
```diff
@@ -5,22 +5,27 @@ PKG_LICENSE=LGPL-2.1
...
+	cmake -DCMAKE_TOOLCHAIN_FILE=$(PKG_BUILDDIR)/xcompile-toolchain.cmake \
+		  -DCHECK=off -DTESTS=0 -DBENCH=0 -DSHLIB=off -Wno-dev $(RELIC_CONFIG_FLAGS) .
 
-$(PKG_BUILDDIR)/Makefile: $(PKG_BUILDDIR)/comp-options.cmake
-	cd "$(PKG_BUILDDIR)" && COMP="$(filter-out -Werror=old-style-definition -Werror=strict-prototypes, $(CFLAGS) ) " cmake -DCMAKE_TOOLCHAIN_FILE=comp-options.cmake -DCHECK=off -DTESTS=0 -DBENCH=0 -DSHLIB=off -Wno-dev $(RELIC_CONFIG_FLAGS) .
-
-CFLAGS += -Wno-gnu-zero-variadic-macro-arguments -Wno-unused-function -Wno-newline-eof
+$(PKG_BUILDDIR)/xcompile-toolchain.cmake: fix_source
+	cd $(PKG_BUILDDIR) && \
+	perl $(RIOTTOOLS)/cmake/generate-xcompile-toolchain.pl > xcompile-toolchain.cmake
```
Comment (cladmi) [inline pkg/relic/Makefile]: Same remarks as the other package with `cd`, I now understand where they come from.

**Comment by cladmi** [comment]
Post review thoughts. In fact, I could fix my `make` remarks in another PR if you want, as you basically just duplicated what was in `relic` in `ccn-lite`.

File: dist/tools/cmake/generate-xcompile-toolchain.pl (hunk 3 above)
Comment (smlng) [reply dist/tools/cmake/generate-xcompile-toolchain.pl]: yeah thought about that, too, but started with using what was already in place with `relic` - I think `sh` would be best.

File: makefiles/vars.inc.mk (hunk 4 above)
Comment (smlng) [reply makefiles/vars.inc.mk]: well that search mostly hits `dist/tools` in comments and readme docs, nevertheless you got a point, that this is sub-optimal and inconsistent. I'll look into it.

File: pkg/ccn-lite/Makefile (hunk 5 above)
Comment (smlng) [reply pkg/ccn-lite/Makefile]: good point

File: pkg/ccn-lite/Makefile (hunk 7 above)
Comment (smlng) [reply pkg/ccn-lite/Makefile]: I'm unsure, because IIRC `cmake` generates its intermediate files where it is called from, which in this case must be `$(PKG_BUILDDIR)/src` otherwise the `make` if above won't work.

**Comment by smlng** [comment]
@cladmi I tried to address you comments, I changed the perl script into a shell script. Your hint with using `@D` didn't work for me, and removing the `&& \` didn't either <- but maybe its this (weird) Makefile scripting syntax. Regarding the `RIOTTOOLS` variable, it comes quite handy here, but I also don't want to break other stuff right now. So adapting other usages throughout RIOT source should be done in followup PRs, or not?

File: makefiles/vars.inc.mk (hunk 4 above)
Comment (smlng) [reply makefiles/vars.inc.mk]: most hits are not actual usages but rather comments, e.g. in README.

**Comment by smlng** [comment]
ping! May I squash?

**Comment by smlng** [comment]
ping again, @cladmi and @kaspar030 IMHO your change requests are addressed.

File: makefiles/vars.inc.mk (hunk 4 above)
Comment (cladmi) [reply makefiles/vars.inc.mk]: There are still a lot, all the python tests use it for example.

git grep dist/tools | grep -v -e README -e doc.txt -e.md | wc -l
176

File: pkg/ccn-lite/Makefile (hunk 7 above)
Comment (cladmi) [reply pkg/ccn-lite/Makefile]: You are right, I forgot cmake was working this way.

File: pkg/relic/Makefile [hunk 11]
```diff
@@ -5,22 +5,28 @@ PKG_LICENSE=LGPL-2.1
...
+TOOLCHAIN_FILE=$(PKG_BUILDDIR)/xcompile-toolchain.cmake
+
 all: $(PKG_BUILDDIR)/Makefile
-	"$(MAKE)" -C $(PKG_BUILDDIR) && \
+	$(MAKE) -C $(PKG_BUILDDIR) && \
 	cp $(PKG_BUILDDIR)/lib/librelic_s.a $(BINDIR)/$(PKG_NAME).a
 
-$(PKG_BUILDDIR)/comp-options.cmake: fix_source
-	cd "$(PKG_BUILDDIR)" && perl $(PKG_DIR)/generate-cmake-xcompile.perl > comp-options.cmake
+$(PKG_BUILDDIR)/Makefile: $(PKG_BUILDDIR)/xcompile-toolchain.cmake
```
Comment (cladmi) [inline pkg/relic/Makefile]: This one should be `$(TOOLCHAIN_FILE)` too.

**Comment by cladmi** [comment]
What is the simple test procedure to test on mac if @kYc0o wants to check ?

**Comment by smlng** [comment]
@cladmi I removed `RIOTTOOLS`, saw you already provided a PR for that -> nice, thanks! If @kYc0o wants to verify: try to compile `examples/ccn-lite` on macOS for `samr21-xpro` linking will fail on master without this PR.

**Comment by smlng** [comment]
I think this is ready to merge now.

File: pkg/ccn-lite/Makefile [hunk 12]
```diff
@@ -7,8 +7,18 @@ PKG_LICENSE=ISC
...
+	$(MAKE) -C $(PKG_BUILDDIR)/src && \
+	cp $(PKG_BUILDDIR)/src/lib/libccnl-riot.a ${BINDIR}/ccn-lite.a
+
+$(PKG_BUILDDIR)/src/Makefile: $(TOOLCHAIN_FILE)
+	cd $(PKG_BUILDDIR)/src && \
+	cmake -DCMAKE_TOOLCHAIN_FILE=$(TOOLCHAIN_FILE) \
+		  -DCCNL_RIOT=1 -DRIOT_CFLAGS="${RIOT_CFLAGS}" -DBUILD_TESTING=OFF .
+
+$(TOOLCHAIN_FILE): git-download
+	$(RIOTBASE)/dist/tools/cmake/generate-xcompile-toolchain.sh > $(TOOLCHAIN_FILE)
```
Comment (kaspar030) [inline pkg/ccn-lite/Makefile]: ```$(RIOTBASE)/dist/tools/cmake/generate-xcompile-toolchain.sh > $@```

File: pkg/relic/Makefile [hunk 13]
```diff
@@ -5,22 +5,28 @@ PKG_LICENSE=LGPL-2.1
...
+	COMP="$(filter-out -Werror=old-style-definition -Werror=strict-prototypes, $(CFLAGS) ) " \
+	cmake -DCMAKE_TOOLCHAIN_FILE=$(TOOLCHAIN_FILE) \
+		  -DCHECK=off -DTESTS=0 -DBENCH=0 -DSHLIB=off -Wno-dev $(RELIC_CONFIG_FLAGS) .
 
-$(PKG_BUILDDIR)/Makefile: $(PKG_BUILDDIR)/comp-options.cmake
-	cd "$(PKG_BUILDDIR)" && COMP="$(filter-out -Werror=old-style-definition -Werror=strict-prototypes, $(CFLAGS) ) " cmake -DCMAKE_TOOLCHAIN_FILE=comp-options.cmake -DCHECK=off -DTESTS=0 -DBENCH=0 -DSHLIB=off -Wno-dev $(RELIC_CONFIG_FLAGS) .
-
-CFLAGS += -Wno-gnu-zero-variadic-macro-arguments -Wno-unused-function -Wno-newline-eof
+$(TOOLCHAIN_FILE): fix_source
+	$(RIOTBASE)/dist/tools/cmake/generate-xcompile-toolchain.sh > $(TOOLCHAIN_FILE)
```
Comment (kaspar030) [inline pkg/relic/Makefile]: same

**Comment by kaspar030** [comment]
I think too this has waited long enough.

@smlng feel free to squash in the "$@".

@cladmi you happy here?

File: pkg/ccn-lite/Makefile (hunk 12 above)
Comment (smlng) [reply pkg/ccn-lite/Makefile]: ah, Makefile magic - kind of :) will add

Comment (smlng) [reply pkg/ccn-lite/Makefile]: mhm, doesn't work though - `$@` is expanded to `/xcompile-toolchain.cmake` only, without pkg build dir set.

Comment (kaspar030) [reply pkg/ccn-lite/Makefile]: so is ```$(TOOLCHAIN_FILE)```? ```$@``` should evaluate to the target name.

Comment (smlng) [reply pkg/ccn-lite/Makefile]: I'm not sure whats going wrong fact is:

```
$(RIOTBASE)/dist/tools/cmake/generate-xcompile-toolchain.sh > $@
```

is expanded to 
```
/path/to/RIOT/dist/tools/cmake/generate-xcompile-toolchain.sh > /xcompile-toolchain.cmake
```

which obviously gives an error, because it is not allow to write to my root dir.

Comment (smlng) [reply pkg/ccn-lite/Makefile]: so the pkgbuilddir is empty at this point

Comment (smlng) [reply pkg/ccn-lite/Makefile]: leaving it as is, works

Comment (kaspar030) [reply pkg/ccn-lite/Makefile]: ok, thanks for trying. Weird that it can match the target but doesn't evaluate $@ afterwards?

Comment (cladmi) [reply pkg/ccn-lite/Makefile]: At that time `PKG_BUILDDIR` is not defined because `pkg.mk` has not been included.
Goals are evaluated immediately not deferred. So the rule is really `/xcompile-toolchain.cmake`.

But the problem was also in the `pkg/relic/Makefile` before, so would require a real fix that could come later.
